### ISSUE TITLE
Issues with the UML model

### ISSUE THREAD

**Issue by heidivanparys** [opengeospatial/WFS_FES#217]
I had a look at the latest UML model (#209) and I have some concerns and a proposal for some modifications.

Original diagram:
![image](https://user-images.githubusercontent.com/11427611/57860143-952d2480-77f4-11e9-8b12-fee7634b9a85.png)

1: 
Issue: Operations are defined on the resource that is returned by the operation.
Proposed modification: A new resource type, e.g. with name OGCAPIFeaturesConformantServer, is added, and all operations are moved to it. Appropriate parameters and return types are added to all operations.

Note that [RFC 3986](https://tools.ietf.org/html/rfc3986) says: "A resource is not necessarily accessible via the Internet", so I would definitely argue that it is ok to add this type OGCAPIFeaturesConformantServer to the model.

2:
Issue: Operations have all the same name ("get"), a name that does not say anything about what the operation actually does or returns, and the name is technology-specific.
Proposed modification: Since all operations are moved to one resource type, they need to have different names. The new names are taken from the OpenAPI definition document in Annex B.2. E.g.: `operationId: getLandingPage`

3:
Issue: The name of a resource type should be singular.
Proposed modification: This seems an indication that something is wrong in the model. Class Collections is removed altogether. What is meant is actually an array of feature collections, see operation describeCollections.

4:
Issue: The name of a resource type should be countable, that means that mass nouns such as "metadata" and "conformance" should be avoided.
Proposed modification: This seems an indication that something is wrong in the model. Therefore, CollectionMetadata is removed altogether. Class Conformance is renamed to ConformanceClass, attribute conformanceClasses renamed to uri, the multiplicity is changed on attribute this attribute and on the end of the association from LandingPage to ConformanceClass.

5:
Issue: As far as I understand the standard, it is not possible to navigate from a feature collection directly to features, one has to go via a page. 
Proposed modification: Therefore, the navigability of that end should be set to non-navigable (visualized with a cross with the appropriate diagram settings). See https://www.uml-diagrams.org/association.html#navigability or the UML standard itself. I am aware of the fact that this is usually not done in model following ISO 19100, however, I think it could clarify this part of the standard. Note that EA seems to have a bug that does not show this cross when the relation type is aggregation. Therefore the relation type is association, but EA allows to change the field "Aggregation" to "shared" for roles also for regular associations.
![image](https://user-images.githubusercontent.com/11427611/57859463-4e8afa80-77f3-11e9-8735-5ce2f4e3301a.png)

6:
Issue: IANA link relation types, which are technology-specific, are present on the diagram as role names.
Proposed modification: Add link relation types as aliases in Enterprise Architect and apply the setting below before exporting the figure.
Another, probably better, option would have been to add those as tagged values. However, Enterprise Architect doesn't allow to visualize tagged values of attributes or operations on diagrams.
![image](https://user-images.githubusercontent.com/11427611/57859736-d1ac5080-77f3-11e9-97ec-954f5fd3fa69.png)

7:
Issue: two relations are present from Page to Page, but it should be only one relation.
Proposed modification: Remove one relation and add the roles "previousPage" and "nextPage", change navigability.

8:
Issue: The standard uses the term feature collection, which is also the preferred term in the terms and definitions.
Proposed modification: Change Collection to FeatureCollection.

New diagram:
![New diagram](https://user-images.githubusercontent.com/11427611/57860222-c86fb380-77f4-11e9-95a2-65200a69f3ba.jpg)

New UML model (extension should be changed to xml): 
[ISO 19168-1 Edition 1 - Proposal.txt](https://github.com/opengeospatial/WFS_FES/files/3187470/ISO.19168-1.Edition.1.-.Proposal.txt)

Notes:

- At first, I also added the paths as aliases, but I felt it became too cluttered. Therefore they are now present in a note. Another, probably better, option would have been to have those only as tagged values on the operations. However, Enterprise Architect doesn't allow to visualize tagged values of attributes or operations on diagrams.
- Types are taken from the ISO 19100 models where possible.
- I think it would be useful to add descriptive relation names to the model, I did an attempt to do so.

**Comment by cportele** [comment]
tl;dr

Many thanks for the thorough review and the suggestions. The UML never felt "right", but I still used it as I am so used to using it to describe things. I see now that it doesn’t work. The intent of the diagram is to illustrate the resources implemented by the API, and their relationships, in a simple way that directly maps to API building blocks. It was not the idea to formally specify the API in UML. As a result, the current diagram cuts some corners, for example, when it gets to the operations. And it includes a (failed) attempt to include aspects like paging. In hindsight, these aspects should not have been part of the diagram. My proposal therefore is to drop the UML from the document and replace it with a non-UML diagram that provides an overview of the resource hierarchy.

---
My responses to the comments/proposals:

1. Operations on the resources that return the operation are intentional. This is what Web APIs do. I also find something like OGCAPIFeaturesConformantServer counterintuitive as this gives the impression of a monolithic web service, not a resource-oriented API.

2. Again, this is intentional as this is a Web API and this reflects the HTTP method(s) on the resource. So, yes, this is HTTP-specific, but for a reason. It is not the idea to develop a conceptual model that is valid independent of the web.

3. Why does the name of the resource type have to be singular? In this case the resource is not an array of something. Beside the array, the resource may have other properties, too. For example, part 2 (CRS support) is expected to add a list of CRSs supported by all collections.

4. If I would use a more specific name for the `/conformance` resource, I would use "ConformanceDeclaration" (actually I have been using "getConformanceDeclaration" as the operationId in some API definition), because this is what it is about. It basically is a list of conformance classes represented by their identifiers. Regarding "metadata", I do not see why it is a bad name for a resource in general.

5. I agree, the current model fails at representing all the details of the API. But that was never the intention and I probably should not have tried to include aspects like paging etc. in the model/diagram in the first place. You approach is more accurate, but I am worried about the level of detail that is starting to appear and that the diagram will be more confusing than helpful to (many) readers.

6. Link relation types are not technology-specific. And like in item 2 above, their use is intentional to simplify the link between the diagram and API.

7. Yes, I agree (except the role names, see item 6).

8. The idea behind the use of just "Collections"/"Collection" was/is the link to the work on an OGC API Common where collections may not be restricted to features. All items in the collections that are in scope of this document are features, but others may add requirements for other resources.

In summary, I see that the use of UML was a bad idea. I used it, because that is a tool/language that I am used to and it seemed "good enough", but your comments show the problems with that approach. The UML model is not meant to be another formal specification of the API building blocks, the diagram was meant as an illustrative overview. It has no normative meaning and should not be understood as a formal specification.

The only way to avoid such misunderstandings is to remove UML from the document. I think there is still value in an illustrative resource overview and the resource hierarchy, but that can be schematic.

If there is interest to reverse-engineer a formal UML specification of OGC API building blocks on a conceptual level (in the sense of ISO 19103), then this should not be documented as part of this document, but rather as a new topic of the OGC abstract specification.

**Comment by cportele** [comment]
20-MAY-2019 Conference call: Agreement to remove the UML. The table http://docs.opengeospatial.org/DRAFTS/17-069.html#tldr should be sufficient for developers.

**Comment by heidivanparys** [comment]
> In this case the resource is not an array of something. Beside the array, the resource may have other properties, too. For example, part 2 (CRS support) is expected to add a list of CRSs supported by all collections.

Ok, so the extra entity is needed, I see.

Regarding the singular (3) & countable (4) issue: the easiest way to see this, is to try to construct proper English sentences from the model, using the names and multiplicities.

![image](https://user-images.githubusercontent.com/11427611/58356145-c9e54f80-7e76-11e9-9e31-b7869fa2d616.png)

A landing page has one collections (of type "Collections"). ❌
A landing page links to one collections. ❌
A collections consists of one or more collections. ❌
A collection metadata must have one identifier. ❌
A collection metadata may have one title. ❌

A landing page has one collection of (feature) collections. ✔️
A landing page links to one collection of (feature) collections.✔️
A collection of (feature) collections consists of one or more (feature) collections.✔️
A (feature) collection must have one identifier.✔️
A (feature) collection may have one title. ✔️

> The only way to avoid such misunderstandings is to remove UML from the document. I think there is still value in an illustrative resource overview and the resource hierarchy, but that can be schematic.

I am not convinced that not using UML at all is the one and only solution. At least the notation of UML is known to many in this domain, as well as well-decribed and standardized (what does a box mean, what does a line mean, what do the numbers mean, etc.). With another type of diagram, you would have to explain the notation, otherwise a diagram/sketch/figure isn't worth much either. Of course the level of a model/diagram should be clearly indicated.

Certain things in the draft standard are still nagging me. My experience is that figures often help sorting out issues by starting a discussion, so let me try again, this time with an alternative, conceptual model (thus high-level and web-independent) of the resources returned by an OGC API Features conformant API.

The focus is
- on the terms and definitions for the resources
- on the names (and reading directions) of the relations between the resources
- the multiplicities on the relation ends

![Conceptual model of the resources in the OGC API Features](https://user-images.githubusercontent.com/11427611/58357819-9ad1dc80-7e7c-11e9-98b8-0a98b0d04fa7.jpg)

As a consequence, the tables in the standard would have to change accordingly.

Resource (current draft) | Resource (proposed) | Path | XML root element
-- | -- | -- | --
Feature   collections metadata | Collection of   feature collections | /collections | core:Collections
Feature   collection metadata | Planned feature   collection | /collections/{collectionId} | core:Collections,   with just one entry for the collection collectionId
Feature   collection | Adhoc feature   collection | /collections/{collectionId}/items   (with   parameters limit, bbox and datetime and parameters   for filtering on feature properties) | core:FeatureCollection
Feature | Feature | /collections/{collectionId}/items/{featureId} | substitutable   for gml:AbstractFeature

**Comment by cportele** [comment]
> Regarding the singular (3) & countable (4) issue: the easiest way to see this, is to try to construct proper English sentences from the model, using the names and multiplicities.

I understand your point. But my point is that the resources that are implemented in the API make sense as they are and the current UML just tried to represent / reverse-engineer this in a UML model for illustration purposes. There should be a 1-to-1 mapping. The role names which are used, however, are not really coming from the API or are just the last step in the path, which is arguably something different. Adding them was probably a bad idea and it makes more sense to skip them like you did in your new model proposal. Another option could have been to use the name of the link relation.

> I am not convinced that not using UML at all is the one and only solution. At least the notation of UML is known to many in this domain, as well as well-decribed and standardized (what does a box mean, what does a line mean, what do the numbers mean, etc.). With another type of diagram, you would have to explain the notation, otherwise a diagram/sketch/figure isn't worth much either. Of course the level of a model/diagram should be clearly indicated.

Yes, in the discussion in the SWG call we also ended up with the result that another type of diagram wouldn't help. And that for developers - the target audience - the table would be sufficient and more informative.

> Certain things in the draft standard are still nagging me. My experience is that figures often help sorting out issues by starting a discussion, so let me try again, this time with an alternative, conceptual model (thus high-level and web-independent) of the resources returned by an OGC API Features conformant API.

I very much appreciate your thoughts and input. This has helped to be more clear in naming and describing the resources. More on this below, but let me add first that I do not think that a conceptual model underpinning the API should be part of this standard. If there will ever be such a model, it would have to be another part of the abstract spec.

Of course, we have to be clear about the resources in the API. I have started to update the overview subclause in a new pull request (plus fixing some errors I noticed while editing) to remove the UML and upate the text. See #220.

In the edits I have not used your naming proposals from the table at the end of the previous comment. The reasons are:

* While "Collection of feature collections" is semantically correct, I find it confusing and do not see what this adds to just "Feature collections"/"Collections", which is more intuitive.
* I do not see that "planned"/"ad-hoc" reflects the distinction between the `/collections/{collectionId}` and `/collections/{collectionId}/items` resource. The separation is that the first resource will (currently) only return information *about* the items in the collection including links to resources related to the collection while the latter resource will return (only) items in the collection. We have used "Collection metadata" for first resource and "Collection" for the second resource, but I agree that this not perfect either as the collection is fully represented by the combination of both and one may argue what is metadata. `items` is provided as a sub-resource, because there is a common need to fetch just the items without all the other information of the collection.

I have not done this yet in the PR, but would the following renaming the resources help?

|Path |Old |New
|---|---|---
|`/collections/{collectionId}` |Feature collection metadata |Feature collection 
|`/collections/{collectionId}/items` |Feature collection |Feature collection items

**Comment by jvanulde** [comment]
+1 on the resource renaming. It's possible that `/collections/{collectionId}` can return something other than metadata.

**Comment by heidivanparys** [comment]
> I very much appreciate your thoughts and input. This has helped to be more clear in naming and describing the resources. 

I am happy to hear that the input is useful.

> Of course, we have to be clear about the resources in the API. I have started to update the overview subclause in a new pull request (plus fixing some errors I noticed while editing) to remove the UML and upate the text. See  #220.

The text is indeed more clear already in that commit.

> I have not done this yet in the PR, but would the following renaming the resources help?
> Path 	Old 	New
> `/collections/{collectionId}` 	Feature collection metadata 	Feature collection
> `/collections/{collectionId}/items` 	Feature collection 	Feature collection items

Yes, that would help.

**Comment by cportele** [comment]
I have updated the PR.

In this I have shortened "Feature collection items" to just "Features".

**Comment by cportele** [comment]
Conclusion of the discussion at ISO/TC 211 meeting:

- There are both views. Most of the developers that we talked to about this do not care about the UML diagram and prefer [table 1](http://docs.opengeospatial.org/DRAFTS/17-069r1.html#tldnr) as an overview, but others find a visual overview very helpful.
- It should not be a conceptual model for the reasons discussed (that would have to be part of a separate specification - in OGC it would be another topic of the abstract specification for the OGC API family). 
- @cportele will make another attempt for a UML model better reflecting the Web API aspects (paths, operations, parameters, link relations, etc), @heidivanparys will review. We will keep this outside of the standard, also that we can improve it later on more easily. First we will add it to the documentation on GitHub and later move it to the Guide.
- The eap/xmi file should be in GitHub (the xmi already is) so that it can be used in EA / tools.
- The updated text in the Overview chapter is an improvement to the current/previous text.

**Comment by cportele** [comment]
@heidivanparys - please have a look, if a diagram like this would help

![image](https://user-images.githubusercontent.com/4245251/58916359-cded5900-8723-11e9-9ad8-3fc832c42a14.png)

Thanks!

**Comment by heidivanparys** [comment]
My comments:

![image](https://user-images.githubusercontent.com/11427611/59098082-a19e2c00-891f-11e9-8bca-6073031f02d6.png)

1. I would like to have the navigability on the association ends back in again (this goes for all relations, not just the one indicated on the annotated figure).
2. All relations should be associations, not compositions. Compositions will likely introduce confusion again, and `Collection` and `Features` together represent the concept "feature collection", as you wrote earlier.
> We have used "Collection metadata" for first resource and "Collection" for the second resource, but I agree that this not perfect either **as the collection is fully represented by the combination of both** and one may argue what is metadata.
3. `collectionId` should be changed back to `id`, as the [schema](https://raw.githubusercontent.com/opengeospatial/WFS_FES/master/core/openapi/schemas/collectionInfo.yaml) specifies it as `id`.
4. `featureId` should probably also be changed back to `id`, when looking at the examples:
```
[ ... ]
"id" : "123",
    "geometry" : {
      "type" : "Polygon",
      "coordinates" : [ ... ]
    }, 
[ ... ]
```
5. The paging stereotype does not really work. See alternative proposal below.
6. Should extent be a derived attribute? I know that the draft says:
> An optional extent that can be used to provide an indication of the spatial and temporal extent of the collection - typically derived from the data.

So from the point of view from the data provider, the extent would indeed be derived from the extent of the featres. But from the point of view from the IT developer using the API, it is not really derived, is it? I can see the reason for adding `Features.collectionId` and `Feature.collectionId` and having them marked as derived, as it can be derived from the path, but that seems to be a different situation.

Suggestions for improvement, now that the abstraction level for the model is clear:
1. Change the UML type of the resources from interface to class, and add stereotype `Resource`.
2. Add assocation names and reading directions. As far as I can see, there are two types of relations:
   1. "includes a link to": should have the relation link type, e.g. as the role name, as has been done already
   2. "contains" : should have no role name at all
3. By following this approach, it seems to become clear that there should be two links of type "includes a link to" from `Features` to itself: one with role name `prev` and one with role name `next`. Together these relations implement the paging concept.
4. Place the `Collections`, `Collection`, `Features` and `Feature` resources underneath each other, in order to emphasize the hierarchy.
5. Maybe a minor thing, but I would suggest to add an attribute with name `...` to `Feature`. From the UML spec:
![image](https://user-images.githubusercontent.com/11427611/59097611-4f104000-891e-11e9-86cc-6c3f7a9b29ec.png)
6. When the model is more finished, create different diagrams that show different aspects of the model, in order to avoid cluttered diagrams:
   1. A diagram with only the resources served by the API (so no data types).
   2. A "context diagram" for `Collection`, with all the data types that it uses.
   3. Diagrams that illustrate possible links to external resources. External resources could e.g. have another color.

![Resources in the Geospatial API for Features - Part 1 Core](https://user-images.githubusercontent.com/11427611/59098066-91864c80-891f-11e9-8193-338e209e73fe.jpg)
![Links to external resources](https://user-images.githubusercontent.com/11427611/59098068-94813d00-891f-11e9-8714-b0c72798b3d8.jpg)

**Comment by cportele** [comment]
Excellent suggestions, thanks @heidivanparys. This looks very good to me. I will follow all of them with the following minor exceptions:

* Support for `prev` is no longer mandatory, so I will not include it.
* I am still struggling  with `id` vs `collectionId`/`featureId` since we have both, one as a member and one as a path parameter. But I think you are correct that the attribute should be `id` to be consistent with the member name. Maybe I will add a note or so.

**Comment by cportele** [comment]
I have updated the pull request #220 to include the UML as discussed in this thread.
